File: scripts/split.py
import os
import re

input_file = os.path.join(os.path.dirname(__file__), '../Youtube_Tools_All_In_One_Optimized.user.js')
out_dir = os.path.join(os.path.dirname(__file__), '../src')

if not os.path.exists(out_dir):
    os.makedirs(out_dir)

with open(input_file, 'r', encoding='utf-8') as f:
    content = f.read()

# Let's split this automatically combining regex and heuristics
# We know the first chunk is the metadata

idx_init = content.find('(function () {')
if idx_init == -1:
    print("Cannot find (function () {")
    exit(1)

meta_chunk = content[:idx_init]
remaining = content[idx_init:]

with open(os.path.join(out_dir, '00_meta.js'), 'w', encoding='utf-8') as f:
    f.write(meta_chunk)

# Find all blocks separated by // ------------------------------
parts = re.split(r'(// ------------------------------\n)', remaining)

# parts[0] is the content before the first // ------------------------------
with open(os.path.join(out_dir, '01_init_and_globals.js'), 'w', encoding='utf-8') as f:
    f.write(parts[0])

# The rest are in parts[1], parts[2], etc...
# parts[1] is the separator itself, parts[2] is the content, parts[3] is separator, parts[4] is content.
counter = 2
for i in range(1, len(parts) - 1, 2):
    separator = parts[i]
    content_chunk = parts[i+1]
    
    # Generate a name based on the first line of content_chunk
    first_line = content_chunk.strip().split('\n')[0].strip()
    clean_name = re.sub(r'[^a-zA-Z0-9]', '_', first_line).strip('_').lower()
    
    # If the chunk is too small or doesn't have a good comment, just use a generic name
    if not clean_name:
        clean_name = f"section_{counter}"
        
    filename = f"{counter:02d}_{clean_name[:40]}.js"
    
    with open(os.path.join(out_dir, filename), 'w', encoding='utf-8') as f:
        f.write(separator + content_chunk)
        
    print(f"Saved {filename}")
    counter += 1

print("Splitting complete based on block separators!")

File: scripts/split.js
const fs = require('fs');
const path = require('path');

const inputFile = path.join(__dirname, '../Youtube_Tools_All_In_One_Optimized.user.js');
const outDir = path.join(__dirname, '../src');

if (!fs.existsSync(outDir)) fs.mkdirSync(outDir);

let content = fs.readFileSync(inputFile, 'utf-8');

// Define exactly where to split the file
// Each section cuts off at the specified 'marker' and writes what was before it to the 'filename'
const markers = [
  { filename: '00_meta.js', marker: '(function () {' },
  { filename: '01_globals.js', marker: '// for translate comments video' },
  { filename: '02_languages.js', marker: 'const waveStyle = \'dinamica\';' },
  { filename: '03_core_utils.js', marker: 'function startDownloadVideoOrAudio(' },
  { filename: '04_download_utils.js', marker: 'const STORAGE = {' },
  { filename: '05_constants.js', marker: '// ------------------------------\r\n// PERF: runtime guards' },
  { filename: '06_runtime.js', marker: '// ------------------------------\r\n// Feature helpers' },
  { filename: '07_helpers.js', marker: '// ------------------------------\r\n// Feature: History' },
  { filename: '08_history.js', marker: '// ------------------------------\r\n// Feature: Show channel name' },
  { filename: '09_shorts_channel.js', marker: '// ------------------------------\r\n// Feature: Show cached rating' },
  { filename: '10_cached_ratings.js', marker: '// ------------------------------\r\n// Feature: Bookmarks' },
  { filename: '11_bookmarks.js', marker: '// ------------------------------\r\n// Feature: Like vs Dislike' },
  { filename: '12_likes_dislikes.js', marker: '// ------------------------------\r\n// YTM Ambient Mode' },
  { filename: '13_ambient.js', marker: 'function isWatchPage() {' },
  { filename: '14_cinema.js', marker: 'function addDownloadBtnAlternative() {' },
  { filename: '15_download_btn.js', marker: 'function initYTTools() {' },
  { filename: '99_init.js', marker: 'null' } // End of file
];

let remainingChunks = content;

markers.forEach((step, i) => {
  if (step.marker === 'null') {
    fs.writeFileSync(path.join(outDir, step.filename), remainingChunks);
    console.log(`Saved ${step.filename}`);
    return;
  }

  // Handle differences in OS line endings
  let index = remainingChunks.indexOf(step.marker);
  if (index === -1) {
    // Try LF instead of CRLF
    index = remainingChunks.indexOf(step.marker.replace(/\r\n/g, '\n'));
  }

  if (index === -1) {
    console.error(`ERROR: Split marker not found for ${step.filename}!`);
    console.error(`Marker: ${step.marker.substring(0, 50)}...`);
    process.exit(1);
  }

  const chunk = remainingChunks.substring(0, index);
  remainingChunks = remainingChunks.substring(index);
  
  fs.writeFileSync(path.join(outDir, step.filename), chunk);
  console.log(`Saved ${step.filename}`);
});

console.log('Splitting complete!');
